### ISSUE TITLE
Lost context by invoke methods from template

### ISSUE THREAD

**Issue by nightire** [NullVoxPopuli/ember-functions-as-helper-polyfill#103]
Take xstate as an example, if we want to invoke the `state.can` method from template will throw an error because `state.can` method need the `state` as the context of the invocation.

```htmlbars
{{@state.can "EVENT_NAME"}}
```

I read the RFC again, it did not mentioned about preserving the invoking context, should we reconsider this or it's the intentional design so that the developers need to handle the context binding by themselves?

I remember we did this once in ember-statechart-component, I had to re-bind the interpreter.send method before we exposes it to outside

**Comment by NullVoxPopuli** [comment]
The template doesn't and (currently) can't know the context a function is supposed to have, so it'll have to be bound before-hand.

Though, I personally think this is an xstate bug where the functions don't have a bound this, and try to rely on JavaScript's "Implicit `this` behavior", which we all know is a bit whacky. lol

**Comment by nightire** [comment]
So in the above example, we can not know that the `@state` is the invoking context?

**Comment by NullVoxPopuli** [comment]
we cannot. It's the same issue as if you were to, in JS, pass the state object to a different class and try to invoke `can`

**Comment by NullVoxPopuli** [comment]
afaik, the template only has a reference to the `can` function, and after `@state.can` has been evaluated, it no longer has a need for `@state`.

I'm sure we _could_ change that, but idk what the implications would be.

If XState were to bind all their methods, than the issue is fixed for everyone.

**Comment by nightire** [comment]
> in JS, pass the state object to a different class and try to invoke can

I don't think so, in the same example, if I invoke this on the JS side, it would be:

```js
invokeCan() {
  this.args.state.can('EVENT_NAME");
}
```

and it will works because `this.args.state` will be the context of this invocation, but `{{@state.can "EVENT_NAME"}}` does not works in the same way, I think it will confuse the developer

> If XState were to bind all their methods, than the issue is fixed for everyone.

xstate defines the `can` method in a very standard way:

https://github.com/statelyai/xstate/blob/main/packages/core/src/State.ts#L359

I think most JS libraries does this, I'm not sure how they should bind all their methods? I'm not sure if this is a bug at its side?

**Comment by NullVoxPopuli** [comment]
> I think most JS libraries does this, I'm not sure how they should bind all their methods? I'm not sure if this is a bug at its side?

This is a very well discussed problem in JavaScript.
Pretty much everyone wants method binding by default.


> I don't think so, in the same example, if I invoke this on the JS side, it would be:

here is a non-ember demonstration of the issue (to eliminate variables and such):

https://jsbin.com/tojigos/edit?html,output

```js
    class State {
      foo = 'my state';
      
      can() {
        return this.foo;
      }
    }
    
    let state = new State();
    
    console.log(state.can());
    
    function evaluateFn(fn) {
      // error: this is undefined
      return fn();
    }
    
    let result = evaluateFn(state.can);

     console.log({ result });
```

the template is pretty much that `evaluateFn` situation. there is no way to get the `this` because it's the function that's passed, and the function isn't bound.
Thus, I think we need to PR to xstate to fix this everywhere.
(and likely any other class-using libraries)

**Comment by nightire** [comment]
@NullVoxPopuli Okay, I understand it clearer, so the template only knows the function itself, but lack of the calling information.

**Comment by NullVoxPopuli** [comment]
yup, exactly

**Comment by NullVoxPopuli** [comment]
I opened: https://github.com/statelyai/xstate/issues/3024
